Test "test_detach_attach_2_data_volumes" fails with AttributeError: 'OCS' object has no attribute 'attachments'

Test_disk_failures ->test_detach_attach_2_data_volumes fails   with AttributeError: 'OCS' object has no attribute 'attachments'


             worker_and_volume['volume'], nodes.detach_volume(worker_and_volume['worker'])
            )

tests/manage/z_cluster/nodes/test_disk_failures.py:127: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
ocs_ci/ocs/platform_nodes.py:485: in detach_volume
    self.aws.detach_volume(volume)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ocs_ci.utility.aws.AWS object at 0x7fd5f97383c8>
volume = <ocs_ci.ocs.resources.ocs.OCS object at 0x7fd5fcff3160>, timeout = 120

    def detach_volume(self, volume, timeout=120):
        """
        Detach volume if attached
    
        Args:
            volume (Volume): The volume to delete
            timeout (int): Timeout in seconds for API calls
    
        Returns:
            Volume: ec2 Volume instance
    
        """
>       if volume.attachments:
E       AttributeError: 'OCS' object has no attribute 'attachments'

ocs_ci/utility/aws.py:351: AttributeError

Below run has more details:
https://ocs4-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/qe-deploy-ocs-cluster/8137/testReport/junit/tests.manage.z_cluster.nodes.test_disk_failures/TestDiskFailures/test_detach_attach_2_data_volumes/

--- a/ocs_ci/ocs/constants.py
+++ b/ocs_ci/ocs/constants.py
@@ -72,6 +72,7 @@ STATUS_BOUND = 'Bound'
 STATUS_RELEASED = 'Released'
 STATUS_COMPLETED = 'Completed'
 STATUS_ERROR = 'Error'
+STATUS_CLBO = 'CrashLoopBackOff'
 
 # NooBaa statuses
 BS_AUTH_FAILED = 'AUTH_FAILED'

--- a/tests/manage/z_cluster/nodes/test_disk_failures.py
+++ b/tests/manage/z_cluster/nodes/test_disk_failures.py
@@ -78,10 +78,10 @@ class TestDiskFailures(ManageTest):
         Detach and attach worker volume
 
         - Detach the data volume from one of the worker nodes
-        - Validate cluster functionality, without checking cluster and Ceph
-          health (as one node volume is detached, the cluster will be unhealthy)
-          by creating resources and running IO
         - Wait for the volumes to be re-attached back to the worker node
+        - Validate cluster functionality, without checking cluster and Ceph
+          health (as one node volume is detached, the cluster will be
+          unhealthy) by creating resources and running IO
         - Restart the node so the volume will get re-mounted
 
         """
@@ -89,9 +89,25 @@ class TestDiskFailures(ManageTest):
         data_volume = nodes.get_data_volumes()[0]
         # Get the worker node according to the volume attachment
         worker = nodes.get_node_by_attached_volume(data_volume)
-
-        # Detach volume (logging is done inside the function)
-        nodes.detach_volume(data_volume, worker)
+        try:
+            # Detach volume (logging is done inside the function)
+            nodes.detach_volume(data_volume, worker)
+        except AWSTimeoutException as e:
+            if "Volume state: in-use" in e:
+                logger.info(
+                    f"Volume {data_volume} re-attached successfully to worker"
+                    f" node {worker}")
+            else:
+                raise
+        else:
+            """
+            Wait for worker volume to be re-attached automatically 
+            to the node
+            """
+            assert nodes.wait_for_volume_attach(data_volume),(
+                f"Volume {data_volume} failed to be re-attached to worker "
+                f"node {worker}"
+            )
 
         # Validate cluster is still functional
         # In case the selected node that its volume disk was detached was the one
@@ -101,11 +117,6 @@ class TestDiskFailures(ManageTest):
         assert wait_for_ct_pod_recovery(), "Ceph tools pod failed to come up on another node"
 
         self.sanity_helpers.create_resources(pvc_factory, pod_factory)
-
-        # Wait for worker volume to be re-attached automatically to the node
-        assert nodes.wait_for_volume_attach(data_volume), (
-            "Volume failed to be re-attached to a worker node"
-        )
 
         # Restart the instance so the volume will get re-mounted
         nodes.restart_nodes([worker])

--- a/tests/manage/z_cluster/nodes/test_disk_failures.py
+++ b/tests/manage/z_cluster/nodes/test_disk_failures.py
@@ -101,10 +101,10 @@ class TestDiskFailures(ManageTest):
                 raise
         else:
             """
-            Wait for worker volume to be re-attached automatically 
+            Wait for worker volume to be re-attached automatically
             to the node
             """
-            assert nodes.wait_for_volume_attach(data_volume),(
+            assert nodes.wait_for_volume_attach(data_volume), (
                 f"Volume {data_volume} failed to be re-attached to worker "
                 f"node {worker}"
             )
@@ -151,19 +151,19 @@ class TestDiskFailures(ManageTest):
             # Detach the volume (logging is done inside the function)
             try:
                 nodes.detach_volume(
-                worker_and_volume['volume'], worker_and_volume['worker']
+                    worker_and_volume['volume'], worker_and_volume['worker']
                 )
             except AWSTimeoutException as e:
                 if "Volume state: in-use" in e:
                     logger.info(
-                    f"Volume {worker_and_volume['volume']} re-attached "
-                    f"successfully to worker "
-                    f"node {worker_and_volume['worker']}")
+                        f"Volume {worker_and_volume['volume']} re-attached "
+                        f"successfully to worker "
+                        f"node {worker_and_volume['worker']}")
                 else:
                     raise
             else:
                 """
-                Wait for worker volume to be re-attached automatically 
+                Wait for worker volume to be re-attached automatically
                 to the node
                 """
                 assert nodes.wait_for_volume_attach(

--- a/tests/manage/z_cluster/nodes/test_disk_failures.py
+++ b/tests/manage/z_cluster/nodes/test_disk_failures.py
@@ -28,6 +28,40 @@ class TestDiskFailures(ManageTest):
     Test class for detach and attach worker volume
 
     """
+
+    def detach_volume_and_wait_for_attach(
+        self, nodes, data_volume, worker_node
+    ):
+        """
+         Detach an EBS volume from an AWS instance and wait for the volume
+         to be re-attached
+
+         Args:
+             node (OCS): The OCS object representing the node
+             data_volume (Volume): The ec2 volume to delete
+             worker_node (OCS): The OCS object of the EC2 instance
+
+         """
+        try:
+            # Detach volume (logging is done inside the function)
+            nodes.detach_volume(data_volume, worker_node)
+        except AWSTimeoutException as e:
+            if "Volume state: in-use" in e:
+                logger.info(
+                    f"Volume {data_volume} re-attached successfully to worker"
+                    f" node {worker_node}")
+            else:
+                raise
+        else:
+            """
+            Wait for worker volume to be re-attached automatically
+            to the node
+            """
+            assert nodes.wait_for_volume_attach(data_volume), (
+                f"Volume {data_volume} failed to be re-attached to worker "
+                f"node {worker_node}"
+            )
+
     @pytest.fixture(autouse=True)
     def teardown(self, request, nodes):
         """
@@ -89,25 +123,9 @@ class TestDiskFailures(ManageTest):
         data_volume = nodes.get_data_volumes()[0]
         # Get the worker node according to the volume attachment
         worker = nodes.get_node_by_attached_volume(data_volume)
-        try:
-            # Detach volume (logging is done inside the function)
-            nodes.detach_volume(data_volume, worker)
-        except AWSTimeoutException as e:
-            if "Volume state: in-use" in e:
-                logger.info(
-                    f"Volume {data_volume} re-attached successfully to worker"
-                    f" node {worker}")
-            else:
-                raise
-        else:
-            """
-            Wait for worker volume to be re-attached automatically
-            to the node
-            """
-            assert nodes.wait_for_volume_attach(data_volume), (
-                f"Volume {data_volume} failed to be re-attached to worker "
-                f"node {worker}"
-            )
+
+        # Detach volume and wait for the volume to attach
+        self.detach_volume_and_wait_for_attach(nodes, data_volume, worker)
 
         # Validate cluster is still functional
         # In case the selected node that its volume disk was detached was the one
@@ -146,32 +164,12 @@ class TestDiskFailures(ManageTest):
             {'worker': nodes.get_node_by_attached_volume(vol), 'volume': vol}
             for vol in data_volumes
         ]
-
         for worker_and_volume in workers_and_volumes:
-            # Detach the volume (logging is done inside the function)
-            try:
-                nodes.detach_volume(
-                    worker_and_volume['volume'], worker_and_volume['worker']
-                )
-            except AWSTimeoutException as e:
-                if "Volume state: in-use" in e:
-                    logger.info(
-                        f"Volume {worker_and_volume['volume']} re-attached "
-                        f"successfully to worker "
-                        f"node {worker_and_volume['worker']}")
-                else:
-                    raise
-            else:
-                """
-                Wait for worker volume to be re-attached automatically
-                to the node
-                """
-                assert nodes.wait_for_volume_attach(
-                    worker_and_volume['volume']
-                ), (f"Volume {worker_and_volume['volume']} failed to be "
-                    f"re-attached to a worker node"
-                    )
-
+            # Detach volume and wait for the volume to attach
+            self.detach_volume_and_wait_for_attach(
+                nodes, worker_and_volume['volume'],
+                worker_and_volume['worker']
+            )
         # Restart the instances so the volume will get re-mounted
         nodes.restart_nodes(
             [worker_and_volume['worker'] for worker_and_volume in workers_and_volumes]

--- a/tests/manage/z_cluster/nodes/test_disk_failures.py
+++ b/tests/manage/z_cluster/nodes/test_disk_failures.py
@@ -55,7 +55,7 @@ class TestDiskFailures(ManageTest):
                     'status'
                 ).get(
                     'containerStatuses'
-                )[0].get('state') == 'CrashLoopBackOff':
+                )[0].get('state') == constants.STATUS_CLBO:
                     node_obj = get_pod_node(pod)
                     nodes.restart_nodes([node_obj])
                     node.wait_for_nodes_status([node_obj.name])

--- a/tests/manage/z_cluster/nodes/test_disk_failures.py
+++ b/tests/manage/z_cluster/nodes/test_disk_failures.py
@@ -14,6 +14,8 @@ from ocs_ci.ocs.resources.pod import (
     get_osd_deployments, get_osd_pods, get_pod_node, get_operator_pods, get_osd_prepare_pods
 )
 from ocs_ci.ocs.resources.ocs import get_job_obj
+from ocs_ci.utility.aws import AWSTimeoutException
+
 
 logger = logging.getLogger(__name__)
 
@@ -45,6 +47,19 @@ class TestDiskFailures(ManageTest):
             if not_ready_nodes:
                 nodes.restart_nodes(not_ready_nodes)
                 node.wait_for_nodes_status()
+
+            # Restart node if the osd stays at CLBO state
+            osd_pods_obj_list = get_osd_pods()
+            for pod in osd_pods_obj_list:
+                if pod.get().get(
+                    'status'
+                ).get(
+                    'containerStatuses'
+                )[0].get('state') == 'CrashLoopBackOff':
+                    node_obj = get_pod_node(pod)
+                    nodes.restart_nodes([node_obj])
+                    node.wait_for_nodes_status([node_obj.name])
+
         request.addfinalizer(finalizer)
 
     @pytest.fixture(autouse=True)
@@ -123,16 +138,28 @@ class TestDiskFailures(ManageTest):
 
         for worker_and_volume in workers_and_volumes:
             # Detach the volume (logging is done inside the function)
-            nodes.detach_volume(
-                worker_and_volume['volume'], nodes.detach_volume(worker_and_volume['worker'])
-            )
-
-        for worker_and_volume in workers_and_volumes:
-            # Wait for worker volume to be re-attached automatically to the node
-            assert nodes.wait_for_volume_attach(worker_and_volume['volume']), (
-                f"Volume {worker_and_volume['volume']} "
-                f"failed to be re-attached to a worker node"
-            )
+            try:
+                nodes.detach_volume(
+                worker_and_volume['volume'], worker_and_volume['worker']
+                )
+            except AWSTimeoutException as e:
+                if "Volume state: in-use" in e:
+                    logger.info(
+                    f"Volume {worker_and_volume['volume']} re-attached "
+                    f"successfully to worker "
+                    f"node {worker_and_volume['worker']}")
+                else:
+                    raise
+            else:
+                """
+                Wait for worker volume to be re-attached automatically 
+                to the node
+                """
+                assert nodes.wait_for_volume_attach(
+                    worker_and_volume['volume']
+                ), (f"Volume {worker_and_volume['volume']} failed to be "
+                    f"re-attached to a worker node"
+                    )
 
         # Restart the instances so the volume will get re-mounted
         nodes.restart_nodes(

